File: CSharp/74-SearchA2DMatrix/74SearchA2DMatrix.cs
// Binary Search

// Time Complexity: O(log(m * n)) (Analyze Complexity Confirmed)
// cut the entire grid in half each time you binary search.
// basically we treat it as one long array

// Space Complexity: O(1) (Analyze Complexity Confirmed)
// just traversing the array that is passed in

public class Solution {
    public bool SearchMatrix(int[][] matrix, int target) {
        int m = matrix.Length;
        int n = matrix[0].Length;
        int left = 0, right = m * n - 1;

        while (left <= right){
            int mid = left + (right - left) / 2;
            int mid_val = matrix[mid/n][mid % n];
            if (mid_val == target) return true;
            else if (mid_val < target) left = mid + 1;
            else right = mid - 1;
        }
        return false;
    }
}

File: Javascript/74-SearchA2DMatrix/74SearchA2DMatrix.js
/* Binary Search
* Time Complexity: O(logr + logc)
* you can binary search via the rows, and then binary search through the columns. 
* this is better than treating it as one big list because that would be O(m log n)
*
* Space Complexity: O(1)
* Just iterating over the passed in matrix. row and c only vars no matter size of input.
*****************************
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    for (let row = 0; row < matrix.length; row++) {
        for (let c = 0; c < matrix[row].length; c++) {
            if (matrix[row][c] == target) {
                return true;
            }
        }
    }
    return false;
} 
